onnxruntime/core/session/inference_session.h

```diff
@@ -305,6 +305,16 @@ class InferenceSession {
                                    std::vector<OrtValue>* p_fetches,
                                    const std::vector<OrtDevice>* p_fetches_device_info = nullptr);
 
+  [[nodiscard]] common::Status Run(const OrtRunOptions* run_options, const char* const* input_names,
+                                   const OrtValue* const* input, size_t input_len,
+                                   const char* const* output_names, size_t output_names_len,
+                                   OrtValue** output);
+
+  [[nodiscard]] common::Status RunAsync(const OrtRunOptions* run_options, const char* const* input_names,
```

Comment:
Please document public methods, including capturing why we're not using gsl::span and a reference for RunOptions here (as that's inconsistent with the other Run methods).

Reply:
The reason for not using gsl::span is to align with the Run(...) defined nearby.
And where the inconsistency from for RunOptions?

Reply:
Can we maybe keep the common code that was adapted between C API types and constraints in onnxruntime_c_api.cc in a helper func so the internal inference_session.h/cc only deals in internal C++ types? it's feels like it might get really confusing to bleed the external API types into the internal class implementation.

Reply:
Yes, this is exactly what I discussed with Randy today evening.

Reply:
The helper function may not be that intuitive - for RunSync, any vectors generated in the helper would be gone when Run is invoked in the thread pool.